File: Yandex/Trainings/S4/Bruteforce/MaxCut/non_recursive_dp_gray_code_2_n.py
def solve():
	n = int(input())
	
	graph = [list(map(int, input().split())) for _ in range(n)]
	
	def get_bit(mask, bit):
		return (mask >> bit) & 1
		
	subsets = (1 << n)
	cuts = [0]
	
	indexes = dict(
		(1 << v, v)
		for v in range(n)
	)
	
	for v in range(n):
		v_graph = graph[v]
		total_sum = sum(v_graph[:v])
				
		pref_subsets = (1 << v)		
		
		cuts.extend(cuts)
		
		v_0 = 0
		prev_subset = 0
		cuts[1 << v] = total_sum
		
		for i in range(1, pref_subsets):
			subset = i ^ (i >> 1)
			
			diff_bit = subset ^ prev_subset
			diff_u = indexes[diff_bit]
			
			prev_subset = subset
			
			w_delta = v_graph[diff_u]
			v_0 += w_delta if get_bit(subset, diff_u) else -w_delta
			
			cuts[subset] += v_0
			cuts[subset | (1 << v)] += total_sum - v_0
		
	ans_cut, ans_subset = max(
		zip(cuts, range(subsets))
	)		

	ans_parts = [get_bit(ans_subset, bit) for bit in range(n)]
	
	print(ans_cut)
	print(*[part + 1 for part in ans_parts])

if __name__ == '__main__':
	solve()

File: Yandex/Trainings/S4/Bruteforce/MaxCut/non_recursive_dp_gray_code_2_n.cpp
#include <bits/stdc++.h>

using namespace std;

using vi = vector<int>;

int get_bit(int mask, int bit) {
	return (mask >> bit) & 1;
}

void solve() {
	int n;
	cin >> n;
	
	vector<vi> graph(n, vi(n));
	for (auto& row : graph) for (auto& edge : row) cin >> edge;
	
	int subsets = (1 << n);
	vi cuts(subsets, 0);
	
	map<int, int> indexes;
	for (int v = 0; v < n; ++v) indexes[1 << v] = v;
	
	for (int v = 0; v < n; ++v) {
		auto& v_graph = graph[v];
		
		int total_sum = 0;
		for (int u = 0; u < v; ++u) total_sum += v_graph[u];
				
		int pref_subsets = (1 << v);
		
		for (int subset = 0; subset < pref_subsets; ++subset) {
			cuts[subset | (1 << v)] = cuts[subset];
		}
		
		int v_0 = 0;
		int prev_subset = 0;
		cuts[1 << v] += total_sum;
		
		for (int i = 1; i < pref_subsets; ++i) {
			int subset = i ^ (i >> 1);
			
			int diff_bit = subset ^ prev_subset;
			int diff_u = indexes[diff_bit];
		
			int w_delta = v_graph[diff_u];
			v_0 += get_bit(subset, diff_u) ? w_delta : -w_delta;
			
			cuts[subset] += v_0;
			cuts[subset | (1 << v)] += total_sum - v_0;
			
			prev_subset = subset;
		}
	}
	
	int max_cut = -1, max_subset = -1;
	for (int subset = 0; subset < subsets; ++subset) {
		int cut = cuts[subset];
		if (max_cut < cut) {
			max_cut = cut;
			max_subset = subset;
		}	
	}
	
	cout << max_cut << "\n";
	for (int v = 0; v < n; ++v, max_subset >>= 1) {
		int part = max_subset & 1;
		cout << part + 1 << " ";
	}
	cout << "\n";
}

int main() {
	solve();
	return 0;
}
